tests/unit_tests/memwipe.cpp

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2017, The Monero Project
+// 
+// All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+// 
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+// 
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "gtest/gtest.h"
+
+#include "misc_log_ex.h"
+#include "common/memwipe.h"
+
+// Probably won't catch the optimized out case, but at least we test
+// it works in the normal case
+static void test(bool wipe)
+{
+  char *foo = (char*)malloc(4);
+  ASSERT_TRUE(foo != NULL);
+  strcpy(foo, "bar");
+  void *bar = wipe ? memwipe(foo, 3) : memset(foo, 0, 3);
+  ASSERT_EQ(foo, bar);
+  free(foo);
+  char *quux = (char*)malloc(4); // same size, just after free, so we're likely to get the same, depending on the allocator
+  if (quux == foo)
```

Comment:
FWIW, even comparing pointers after freeing one of them is UB. I think you're conflating memory management with zeroing here. While we care that the free'd memory has been zeroed out, that's not something that C allows you to test. That said, we don't need to worry about malloc/free here... this is sufficient:

```
  char foo[] = "foo";
  volatile char *wipe = memwipe(foo, sizeof(foo));
  ASSERT_EQ(wipe, foo);

  // Use a volatile ptr so the compiler cannot elide the checks:
  volatile char *vfoo = foo;
  ASSERT_EQ(vfoo[0], '0');
  ASSERT_EQ(vfoo[1], '0');
  ASSERT_EQ(vfoo[2], '0');
  ASSERT_EQ(vfoo[3], '0');
```

Reply:
Are you sure comparing pointers after free is UB ?

Your example defeats the point of the test because those accesses will likely cause any optimization pass to not remove the memset in the first place.

Reply:
I did find references saying it is indeed UB. Bastards. Oh well.

Reply:
Wearing my compiler engineer hat, yes, it is UB :)

Short of inspecting the assembly, I do not know of any way to check for the presence of the optimization barrier in memwipe's implementation.

Reply:
Fair enough. I'll leave it as its latest form, as it will only access through a valid pointer. If the foo pointer is somehow changed, it'll go to the "cannot test" branch, so we can't lose I think.

Reply:
Oh, codesourcery... thank you all for your past work as a whole ^_^

Reply:
We lose when the compiler realizes that the pointer test is UB, and decides that it's always going to take the false branch. std::launder might help here, but we won't get that until c++17 :/

Reply:
Ah, good point :/

Reply:
Though we still don't lose, we just fail to win ^_^

Reply:
Or if it shoots out nasal demons, sends all your btc to `1bitcoineateraddressdontsendf59kue`, and formats your hdd :P

You really can't make any assumptions about what happens in the presence of UB.

Reply:
How about this. We save foo into an intptr_t before the free. Then we compare later. Still UB ?

```
static void test(bool wipe)
{
  char *foo = (char*)malloc(4);
  ASSERT_TRUE(foo != NULL);
  strcpy(foo, "bar");
  void *bar = wipe ? memwipe(foo, 3) : memset(foo, 0, 3);
  ASSERT_EQ(foo, bar);
  intptr_t foop = (intptr_t)foo;
  free(foo);
  char *quux = (char*)malloc(4); // same size, just after free, so we're likely to get the same, depending on the allocator
  if ((intptr_t)quux == foop)
  {
    MDEBUG(std::hex << std::setw(8) << std::setfill('0') << *(uint32_t*)quux);
    if (wipe) ASSERT_TRUE(!memcmp(quux, "\0\0\0", 3));
  }
  else MWARNING("We did not get the same location, cannot check");
  free(quux);
}
```

Reply:
I believe this still depends on UB. The compiler can see right through the fact that you're trying to catch malloc giving you the same memory again.